Add support for VirtualMachineOperandArray

charliegracie opened the pull request:
Not all Virtual Machines have an Operand Stack. Another common construct
is to have a fixed sized array of "registers". The new
VirtualMachineOperandArray is very similar to the VMOS but it does not
allow for growth and it has to deal with NULL entries as not all entries
may have a value.

The VirtualMachineOperandArray::Move API requires that a copy of the
IlValue stored at srcIndex be stored at dstIndex. To accomplish this I
added a new API to IlBuilder call CopyValue.

I have added tests to verify the all of the functionality.

Signed-off-by: Charlie Gracie <charlie.gracie@gmail.com>

mstoodle commented on compiler/ilgen/VirtualMachineOperandArray.cpp:
```diff
@@ -0,0 +1,183 @@
+      if (element != NULL)
+         {
+         b->StoreAt(
+         b->       IndexAt(pElement,
```
not standard indentation for compiler component files. I know one day we'll change it, but for now please keep it consistent even if more spaces makes the tree structure a tad more readable.

mstoodle commented on compiler/ilgen/VirtualMachineOperandArray.cpp:
```diff
@@ -0,0 +1,183 @@
+   {
+   VirtualMachineOperandArray *other = (VirtualMachineOperandArray *)o;
+   TR_ASSERT(_numberOfElements == other->_numberOfElements, "array are not same size");
+   TR::BytecodeBuilder *builder = (TR::BytecodeBuilder *)b;
```
i don't see `builder` being used anywhere, so please remove it.

mstoodle commented on compiler/ilgen/VirtualMachineOperandArray.hpp:
```diff
@@ -0,0 +1,136 @@
+ *   expressions are stored in the TR::IlValue's for the state being merged
+ *   *to*. So the purpose of MergeInto() is to store the values of the current
+ *   state into the same variables as in the "other" state.
+ *
```
should add UpdateArray() into this list

mstoodle commented on jitbuilder/release/src/OperandArrayTests.hpp:
```diff
@@ -0,0 +1,61 @@
+#include "ilgen/MethodBuilder.hpp"
+
+#define ARRAYVALUEILTYPE	Int32
+#define	ARRAYVALUETYPE		int32_t
```
stray tabs here :)

mstoodle commented on compiler/ilgen/IlBuilder.hpp:
```diff
@@ -142,6 +142,8 @@ class IlBuilder : public TR::IlInjector
    // create a new local value (temporary variable)
    TR::IlValue *NewValue(TR::IlType *dt);
 
+   TR::IlValue *CopyValue(TR::IlValue *value);
```
given the API, I would be tempted to call it "Copy" rather than "CopyValue" .... I know it's kinda like NewValue, but I think of it more like Load, Store, Add, Sub which also could (but by choice did not) get a Value suffix.

I could be argued out of it, but just think of all the characters you will no longer have to type :) .

mstoodle commented on jitbuilder/release/src/OperandArrayTests.cpp:
```diff
@@ -0,0 +1,551 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2016, 2016
```
should be modified in 2017

mstoodle commented on jitbuilder/release/src/OperandArrayTests.cpp:
```diff
@@ -0,0 +1,551 @@
+bool
+OperandArrayTestMethod::testArray(TR::BytecodeBuilder *builder, bool useEqual)
+   {
+   //VMOA      [null,null,null,...,null]
```
I like this style, makes it very clear!

mstoodle commented on jitbuilder/release/src/OperandArrayTests.hpp:
```diff
@@ -0,0 +1,61 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2016, 2016
```
should be modified in 2017

mstoodle commented on jitbuilder/build/rules/common.mk:
```diff
@@ -108,6 +108,9 @@ $(RELEASE_INCLUDE)/$(JIT_OMR_DIRTY_DIR)/ilgen/VirtualMachineRegister.hpp: $(FIXE
 $(RELEASE_INCLUDE)/$(JIT_OMR_DIRTY_DIR)/ilgen/VirtualMachineRegisterInStruct.hpp: $(FIXED_SRCBASE)/$(JIT_OMR_DIRTY_DIR)/ilgen/VirtualMachineRegisterInStruct.hpp $(RELEASE_INCLUDE)/$(JIT_OMR_DIRTY_DIR)/ilgen
 	cp $< $@ || cp $< $@
 
+$(RELEASE_INCLUDE)/$(JIT_OMR_DIRTY_DIR)/ilgen/VirtualMachineOperandArray.hpp: $(FIXED_SRCBASE)/$(JIT_OMR_DIRTY_DIR)/ilgen/VirtualMachineOperandArray.hpp $(RELEASE_INCLUDE)/$(JIT_OMR_DIRTY_DIR)/ilgen
```
please update the copyright header in this file

charliegracie commented:
I believe I addressed all of the issues. 

I made a change to `VirtualMachineOperandArray::MergeInto`.  It now asserts that if the destination array has a NULL value at a given index that the source array is also NULL.

mstoodle commented on jitbuilder/release/src/Simple.cpp:
```diff
@@ -89,7 +89,8 @@ SimpleMethod::buildIL()
    Return(
       Add(
-         Load("value"),
+         Copy(
```
Sorry, Charlie, I just realized that we probably shouldn't be mucking with Simple.cpp because the jitbuilder blog articles (and several of the presentations) reference it and describes it in detail. So I'd prefer not to just add Copy here (though it didn't register with me in my first review).

Given you have OperandArrayTests which presumably exercise Copy, I'm not sure we *really* need it in the code samples.

charliegracie commented:
I updated it accordingly

mstoodle commented on compiler/ilgen/VirtualMachineOperandArray.cpp:
```diff
@@ -0,0 +1,181 @@
+      {
+      if (NULL == other->_values[i])
+         {
+         TR_ASSERT(_values[i] == NULL, "if an element in the dest array at index is NULL it has to be NULL in the src array");
```
This assert establishes a "static" rule that all operands uses must be reached by an earlier definition on every incoming path. I think this is the only reasonable choice, but I wanted to call out that this assertion codifies it (but projects that leverage VirtualMachineOperandArray, anyway :) ).
